src/cloudWatchLogs/changeLogSearch.ts

```diff
@@ -0,0 +1,79 @@
+/*!
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+import * as telemetry from '../shared/telemetry/telemetry'
+import * as vscode from 'vscode'
+import { showInputBox } from '../shared/ui/inputPrompter'
+import { isLogStream } from './cloudWatchLogsUtils'
+import { prepareDocument } from './commands/searchLogGroup'
+import { getActiveUri } from './document/logStreamDocumentProvider'
+import { CloudWatchLogsData, filterLogEventsFromUriComponents, LogStreamRegistry } from './registry/logStreamRegistry'
+import { isViewAllEvents, TimeFilterResponse, TimeFilterSubmenu } from './timeFilterSubmenu'
+
+type ChangeableParam = 'filterPattern' | 'timeFilter'
+
+export async function getNewData(
+    registry: LogStreamRegistry,
+    param: ChangeableParam,
+    oldUri: vscode.Uri
+): Promise<CloudWatchLogsData | undefined> {
+    const oldData = registry.getLogData(oldUri) as CloudWatchLogsData
+    const newData: CloudWatchLogsData = {
+        ...oldData,
+        data: [],
+        next: undefined,
+        previous: undefined,
+    }
+    let newPattern: string | undefined
+    let newTimeRange: TimeFilterResponse | undefined
+
+    switch (param) {
+        case 'filterPattern':
+            newPattern = await showInputBox({
+                title: isLogStream(oldUri) ? 'Filter Log Stream Results' : 'Log Group Keyword Search',
+                placeholder: oldData.parameters.filterPattern ?? 'Enter Text Here',
+            })
+            if (newPattern === undefined) {
+                return
+            }
+            newData.parameters.filterPattern = newPattern
+
+            break
+
+        case 'timeFilter':
+            newTimeRange = (await new TimeFilterSubmenu().prompt()) as TimeFilterResponse
+            if (newTimeRange === undefined) {
+                return
+            }
+            newData.parameters.startTime = isViewAllEvents(newTimeRange) ? undefined : newTimeRange.start
+            newData.parameters.endTime = isViewAllEvents(newTimeRange) ? undefined : newTimeRange.end
+
+            break
+    }
+    if (newData.parameters.streamName) {
+        newData.retrieveLogsFunction = filterLogEventsFromUriComponents
+        newData.parameters.streamNameOptions = [newData.parameters.streamName]
+        newData.parameters.streamName = undefined
+    }
+
+    return newData
+}
+
+export async function changeLogSearchParams(registry: LogStreamRegistry, param: ChangeableParam): Promise<void> {
+    let result: telemetry.Result = 'Succeeded'
+
+    const oldUri = getActiveUri(registry)
+    const newData = await getNewData(registry, param, oldUri)
```

Comment:
Every single `await` is a place where race conditions can occur. JS is a single-threaded language, but multiple 'tasks' can still be executing concurrently.

I don't think we need to worry about this here though. Looks like every async call is pretty clean.

Reply:
When should I worry about this? If the `await` calls begin to depend on each other more?

Reply:
> If the await calls begin to depend on each other more?

Yes. To recognize it, we have to think about all those assumptions we're making that might be violated in a concurrent scenario. Changing state is generally something you want to look out for.

Reply:
Always worry :) but in practice it depends on what the stakes are.

Similar to any filesystem or network call: the data could change just before or just after you read it. In a callback/promise-based system, there is an additional risk of the call getting deferred.

There is a "lock" concept for when you want some sort of atomicity: https://github.com/aws/aws-toolkit-vscode/blob/420cd0158f8a1b9b5e58c531aae1e4fe6e9f0baa/src/cloudWatchLogs/commands/addLogEvents.ts#L32